fat/fat2/transactionio_test.go

```diff
@@ -0,0 +1,120 @@
+}, {
+	Name: "valid (allow zero value)",
+	JSON: `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q","type":"PEG","amount":0}`,
+}, {
+	Name: "valid (type, amount)",
+	JSON: `{"type":"PEG","amount":50}`,
+}, {
+	Name: "valid (address, amount)",
+	JSON: `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q","amount":50}`,
+}, {
+	Name: "valid (address, type)",
+	JSON: `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q","type":"PEG"}`,
+}, {
+	Name:  "invalid (address)",
+	Error: "checksum error",
+	JSON:  `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6Rbad","type":"PEG","amount":50}`,
+}, {
+	Name:  "invalid (negative amount)",
+	Error: "cannot unmarshal number",
+	JSON:  `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q","type":"PEG","amount":-50}`,
+}, {
+	Name:  "invalid (non-string type)",
+	Error: "cannot unmarshal number",
+	JSON:  `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q","type":50,"amount":50}`,
+}}
+
+func TestTransactionIO_UnmarshalJSON(t *testing.T) {
+	for _, test := range transactionIOUnmarshalJSONTests {
+		t.Run(test.Name, func(t *testing.T) {
+			assert := assert.New(t)
+			var txio TransactionIO
+			err := json.Unmarshal([]byte(test.JSON), &txio)
+			if len(test.Error) != 0 {
+				if assert.NotNil(err) {
+					assert.Contains(err.Error(), test.Error)
+				}
+				return
+			}
+			require.NoError(t, err)
+		})
+	}
+}
+
+var transactionIOValidationTests = []struct {
+	Name  string
+	Error string
+	JSON  string
+}{{
+	Name: "valid (address, type, amount)",
+	JSON: `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q","type":"PEG","amount":50}`,
+}, {
+	Name: "valid (address, type)",
+	JSON: `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q","type":"PEG"}`,
+}, {
+	Name: "valid (address, amount)",
+	JSON: `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q","amount":50}`,
+}, {
+	Name: "valid (type, amount)",
+	JSON: `{"type":"PEG","amount":50}`,
+}, {
+	Name: "valid (address)",
+	JSON: `{"address":"FA2jK2HcLnRdS94dEcU27rF3meoJfpUcZPSinpb7AwQvPRY6RL1Q"}`,
+}, {
+	Name: "valid (type)",
+	JSON: `{"type":"PEG"}`,
+}, {
+	Name: "valid (amount)",
+	JSON: `{"amount":50}`,
+}, {
+	Name:  "invalid (empty)",
+	Error: "at least one non-nil field required",
+	JSON:  `{}`,
+}, {
+	Name:  "invalid (wrong fields)",
+	Error: "at least one non-nil field required",
+	JSON:  `{"wrong":"fields"}`,
+}, {
+	Name:  "invalid (unknown asset type)",
+	Error: "invalid token type",
+	JSON:  `{"type":"UNKNOWN"}`,
+}}
+
+func TestTransactionIO_Valid(t *testing.T) {
+	for _, test := range transactionIOValidationTests {
+		t.Run(test.Name, func(t *testing.T) {
+			assert := assert.New(t)
+			var txio TransactionIO
+			err := json.Unmarshal([]byte(test.JSON), &txio)
+			assert.Nilf(err, "JSON Unmarshal raised an unexpected error")
+			err = txio.Valid()
+			if len(test.Error) != 0 {
+				if assert.NotNil(err) {
```

Comment:
Sounds good, was copying your usage of `assert.Contains(err.Error(), test.Error)` in some fat-0 tests and added  the nil check because I noticed it panics when a test unexepectedly has no error.

Will change to `EqualError` and will change all my error strings to be the full error rather than a substring that `Contains` was able to work with